src/ApiService/ApiService/Program.cs

```diff
@@ -58,6 +58,67 @@ public class LoggingMiddleware : IFunctionsWorkerMiddleware {
         }
     }
 
+    /// <summary>
+    /// Represents a middleware that can optionally perform strict version checking based on data sent in request headers.
+    /// </summary>
+    public class VersionCheckingMiddleware : IFunctionsWorkerMiddleware {
+        private const string CliVersionHeader = "Cli-Version";
+        private const string StrictVersionHeader = "Strict-Version";
+        private readonly Version _oneFuzzServiceVersion;
+        private readonly IRequestHandling _requestHandling;
+
+        /// <summary>
+        /// Initializes an instance of <see cref="VersionCheckingMiddleware"/> with the provided config and request handling objects.
+        /// </summary>
+        /// <param name="config">The service config containing the service version.</param>
+        /// <param name="requestHandling">The request handling object to create HTTP responses with.</param>
+        public VersionCheckingMiddleware(IServiceConfig config, IRequestHandling requestHandling) {
+            _oneFuzzServiceVersion = Version.Parse(config.OneFuzzVersion);
+            _requestHandling = requestHandling;
+        }
+
+        public Error? CheckCliVersion(Azure.Functions.Worker.Http.HttpHeadersCollection headers) {
+            var doStrictVersionCheck =
+                headers.TryGetValues(StrictVersionHeader, out var strictVersion)
+                && strictVersion?.FirstOrDefault()?.Equals("true", StringComparison.InvariantCultureIgnoreCase) == true; // "== true" necessary here to avoid implicit null -> bool casting
+
+            if (doStrictVersionCheck) {
+                if (!headers.TryGetValues(CliVersionHeader, out var cliVersion)) {
+                    return Error.Create(ErrorCode.INVALID_REQUEST, $"'{StrictVersionHeader}' is set to true without a corresponding '{CliVersionHeader}' header");
+                }
+                if (!Version.TryParse(cliVersion?.FirstOrDefault() ?? "", out var version)) {
```

Comment:
`System.Version` doesn't support our dev versioning scheme. 

>... All defined components must be integers greater than or equal to 0. ...
https://learn.microsoft.com/en-us/dotnet/api/system.version?view=net-7.0#remarks

Take a look at this code sample I made with the current version in my dev deployment. https://dotnetfiddle.net/44IZLZ

We need to use a library that supports semver for version parsing.

Reply:
Thanks for catching that! I went for [Semver](https://github.com/maxhauser/semver) for parsing to replace System.Version.